### PR TITLE
Time reasoning

### PR DESCRIPTION
**Pull request by Pialex99**
Add time reasoning functionality.
- The error reporting is a bit hacky but it works.
- Does not work on pure functions and closures.
- Quadratic runtime does not work because z3 can't handle it out of the box.

### CODE REVIEW

File: prusti-common/src/vir/optimizations/functions/simplifier.rs [hunk 1]
```diff
@@ -46,134 +50,118 @@ impl ExprSimplifier {
...
-                        ..
+                        position: inner_pos,
                     }),
                 position: pos,
             }) => ast::Expr::Const(ast::ConstExpr {
                 value: ast::Const::Bool(!b),
                 position: pos,
-            }),
+            })
+            .set_default_pos(inner_pos),
```
Comment (Aurel300) [inline prusti-common/src/vir/optimizations/functions/simplifier.rs:60]: What does this do? If it's a constant Boolean, there shouldn't be any more position to override with `set_default_pos`, right?

File: prusti-common/src/vir/optimizations/functions/simplifier.rs [hunk 2]
```diff
@@ -46,134 +50,118 @@ impl ExprSimplifier {
...
-                ..
-            }) => {
-                if b {
-                    conjunct
-                } else {
-                    false.into()
-                }
+                position: pos,
+            }) => if b {
+                conjunct
```
Comment (Aurel300) [inline prusti-common/src/vir/optimizations/functions/simplifier.rs:98]: @vakaras If we inline a `dummy` trigger function body (with the body `true`) and also simplify, this would change semantics of quantifiers. Do you know in what order the inlining is performed?

File: prusti-common/src/vir/optimizations/functions/simplifier.rs [hunk 3]
```diff
@@ -46,134 +50,118 @@ impl ExprSimplifier {
...
+            } else {
+                Into::<ast::Expr>::into(false).set_pos(inner_pos)
             }
+            .set_default_pos(pos)
+            .set_default_pos(inner_pos),
```
Comment (Aurel300) [inline prusti-common/src/vir/optimizations/functions/simplifier.rs]: One of these should not need to be here.

File: prusti-common/src/vir/optimizations/functions/simplifier.rs [hunk 4]
```diff
@@ -46,134 +50,118 @@ impl ExprSimplifier {
...
-                    disjunct
-                }
+                position: pos,
+            }) => if b {
+                Into::<ast::Expr>::into(true).set_pos(inner_pos)
+            } else {
+                disjunct
             }
+            .set_default_pos(pos)
+            .set_default_pos(inner_pos),
```
Comment (Aurel300) [inline prusti-common/src/vir/optimizations/functions/simplifier.rs]: Same.

File: prusti-common/src/vir/optimizations/functions/simplifier.rs [hunk 5]
```diff
@@ -46,134 +50,118 @@ impl ExprSimplifier {
...
-                    true.into()
-                }
+                position: pos,
+            }) => if b {
+                body
+            } else {
+                Into::<ast::Expr>::into(true).set_pos(inner_pos)
             }
+            .set_default_pos(inner_pos)
+            .set_default_pos(pos),
```
Comment (Aurel300) [inline prusti-common/src/vir/optimizations/functions/simplifier.rs]: Same.

File: prusti-common/src/vir/optimizations/methods/simplifier.rs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use log::debug;
+use std::mem;
+use vir::polymorphic::*;
+
+use crate::vir::optimizations::functions::Simplifier;
+
+/// This optimization transformes ors were the RHS is impure into implications.
```
Comment (Aurel300) [inline prusti-common/src/vir/optimizations/methods/simplifier.rs]:
```suggestion
/// This optimization transforms ors were the RHS is impure into implications.
```

File: prusti-common/src/vir/optimizations/predicates/delete_unused_predicates.rs [hunk 7]
```diff
@@ -7,18 +7,20 @@
...
-    utils::{walk_functions, walk_methods},
+    utils::{walk_builtin_methods, walk_functions, walk_methods},
 };
 use fxhash::{FxHashMap, FxHashSet};
 use log::debug;
 
 fn collect_info_from_methods_and_functions(
     methods: &[CfgMethod],
     functions: &[Function],
+    buildin_methods: &[BodylessMethod],
```
Comment (Aurel300) [inline prusti-common/src/vir/optimizations/predicates/delete_unused_predicates.rs]: `builtin_methods` (also later in `delete_unused_predicates`).

File: prusti-tests/tests/typecheck/fail/time-resources/condition.rs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+// compile-flags: -Ptime_reasoning=true
+
+use prusti_contracts::*;
+
+// ignore-test: prusti-constract functions used outside specification check not yet implemented
```
Comment (Aurel300) [inline prusti-tests/tests/typecheck/fail/time-resources/condition.rs:5]: I'm not sure about "not yet" here: we don't want to implement something like this. Many specification functions don't make sense in executable code, and branching on `time_credits` or `receipts` is also such an example. I would remove this test.

File: prusti-common/src/vir/to_viper.rs [hunk 9]
```diff
@@ -402,6 +402,15 @@ impl<'v> ToViper<'v, viper::Expr<'v>> for Expr {
                     perm.to_viper(context, ast),
                     pos.to_viper(context, ast),
                 ),
+            Expr::ResourceAccessPredicate(ref resource_name, ref amount, ref pos) => {
+                // Resource are encoded as a predicate with no arguments
```
Comment (Aurel300) [inline prusti-common/src/vir/to_viper.rs]:
```suggestion
                // resources are encoded as predicates with no arguments
```

File: prusti-common/src/vir/to_viper.rs [hunk 10]
```diff
@@ -402,6 +402,15 @@ impl<'v> ToViper<'v, viper::Expr<'v>> for Expr {
...
+                ast.predicate_access_predicate_with_pos(
+                    ast.predicate_access(&[], resource_name),
+                    // ast.int_perm_mul(amount.to_viper(context, ast), ast.full_perm()),
```
Comment (Aurel300) [inline prusti-common/src/vir/to_viper.rs]:
```suggestion
```

File: prusti-utils/src/config.rs [hunk 11]
```diff
@@ -364,6 +368,11 @@ pub fn check_panics() -> bool {
     read_setting("check_panics")
 }
 
+/// When enabled, the time reasoning constrains are verified. Otherwise they are ignored.
```
Comment (Aurel300) [inline prusti-utils/src/config.rs:370]: Please also update `flags.md` accordingly (for the [flags](https://viperproject.github.io/prusti-dev/dev-guide/config/flags.html) page in the dev guide). I would also add the comment that this is an experimental feature for the time being.

File: prusti-viper/src/encoder/errors/error_manager.rs [hunk 12]
```diff
@@ -180,6 +177,10 @@ pub enum ErrorCtxt {
     /// The state that fold-unfold algorithm deduced as unreachable, is actually
     /// reachable.
     UnreachableFoldingState,
+    // There is not enough time credtis to execute this code.
```
Comment (Aurel300) [inline prusti-viper/src/encoder/errors/error_manager.rs]:
```suggestion
    // There is not enough time credits to execute this code.
```

File: prusti-viper/src/encoder/errors/error_manager.rs [hunk 13]
```diff
@@ -322,32 +341,30 @@ impl<'tcx> ErrorManager<'tcx> {
...
+                Some(ref pos_id) => PrustiError::internal(
+                    format!(
+                        "unregistered verification error: [{}; {}] {}",
+                        ver_error.full_id, pos_id, ver_error.message
+                    ),
+                    error_span,
+                )
+                .set_help(
+                    "This could be caused by too small assertion timeout. \
                         Try increasing it by setting the configuration parameter \
```
Comment (Aurel300) [inline prusti-viper/src/encoder/errors/error_manager.rs]: I assume this was an automated reformat, but please align the extra lines with the first one. Same on L365-366.

File: prusti-viper/src/encoder/errors/error_manager.rs [hunk 14]
```diff
@@ -698,6 +715,35 @@ impl<'tcx> ErrorManager<'tcx> {
...
                 )
             }
+            ("exhale.failed:insufficient.permission" | "assert.failed:insufficient.permission", ErrorCtxt::NotEnoughTimeCredits) => {
+                PrustiError::verification("Not enough time credits.".to_string(), error_span)
+            }
+            ("exhale.failed:insufficient.permission" | "assert.failed:insufficient.permission", ErrorCtxt::NotEnoughTimeReceipts) => {
+                PrustiError::verification("Not enough time receipts.".to_string(), error_span)
+            }
+
+            // TODO: this is a temporary fix for detecting time credits/receipts errors
```
Comment (Aurel300) [inline prusti-viper/src/encoder/errors/error_manager.rs]: Please add a more detailed description to the TODO. We might know (roughly) what the problem is now, but we might forget :)

File: prusti-viper/src/encoder/mir/pure/specifications/interface.rs [hunk 15]
```diff
@@ -255,6 +258,19 @@ impl<'v, 'tcx: 'v> SpecificationEncoderInterface<'tcx> for crate::encoder::Encod
...
+                if config::time_reasoning() {
+                    Ok(encode_time_specifications(
+                        self,
+                        &encoded_args[0],
+                        vir_poly::ResourceType::from_function_name(fn_name),
+                        parent_def_id,
+                        span,
+                    ))
+                } else {
+                    Ok(true.into())
```
Comment (Aurel300) [inline prusti-viper/src/encoder/mir/pure/specifications/interface.rs]: For the time being, trying to use these two functions when the flag is enabled should be an error. If this isn't already detected when parsing specifications, then raise an incorrect specification error here.

File: vir/defs/polymorphic/ast/expr.rs [hunk 16]
```diff
@@ -617,7 +633,10 @@ impl Expr {
 
     pub fn is_only_permissions(&self) -> bool {
         match self {
-            Expr::PredicateAccessPredicate(..) | Expr::FieldAccessPredicate(..) => true,
+            Expr::PredicateAccessPredicate(..)
+            // TODO check this not sure
```
Comment (Aurel300) [inline vir/defs/polymorphic/ast/expr.rs]: Assuming this is used for e.g. purification, then setting this to `true` looks correct. For now we are using resources to verify time anyway.

File: vir/src/legacy/ast/expr.rs [hunk 17]
```diff
@@ -773,7 +785,10 @@ impl Expr {
 
     pub fn is_only_permissions(&self) -> bool {
         match self {
-            Expr::PredicateAccessPredicate(..) | Expr::FieldAccessPredicate(..) => true,
+            Expr::PredicateAccessPredicate(..)
+            // TODO check this not sure
```
Comment (Aurel300) [inline vir/src/legacy/ast/expr.rs]: (Same comment as before.)

File: vir/defs/polymorphic/ast/common.rs [hunk 18]
```diff
@@ -489,6 +489,42 @@ impl Type {
...
+    }
+}
+
+impl ResourceType {
+    pub fn from_function_name(function_name: &str) -> ResourceType {
+        match function_name {
+            "prusti_contracts::time_credits" => ResourceType::TimeCredits,
+            "prusti_contracts::time_receipts" => ResourceType::TimeReceipts,
+            _ => unreachable!(
+                "Fonction name: {} does not match any resource type.",
```
Comment (Aurel300) [inline vir/defs/polymorphic/ast/common.rs]:
```suggestion
                "Function name: {} does not match any resource type.",
```

File: prusti-viper/src/encoder/foldunfold/requirements.rs [hunk 19]
```diff
@@ -91,13 +91,17 @@ impl RequiredStmtPermissionsGetter for vir::Stmt {
                 ref targets,
                 ..
             }) => {
-                // Preconditions and postconditions are empty
-                assert!(arguments.is_empty());
+                // Preconditions and postconditions are empty or does not require permissions
```
Comment (Aurel300) [inline prusti-viper/src/encoder/foldunfold/requirements.rs]:
```suggestion
                // Preconditions and postconditions are empty or do not require permissions
```

File: prusti-viper/src/encoder/foldunfold/requirements.rs [hunk 20]
```diff
@@ -91,13 +91,17 @@ impl RequiredStmtPermissionsGetter for vir::Stmt {
...
+                let arguments_perm = arguments
+                    .iter()
+                    .map(|expr| expr.get_required_stmt_permissions(predicates).into_iter())
```
Comment (Aurel300) [inline prusti-viper/src/encoder/foldunfold/requirements.rs]: Why this change? This only computes permissions for arguments, not pre-/postconditions. I guess for us that's what we want, because the time resource permissions should not interact with fold/unfold anyway, but the comment should make this clear.

File: prusti-viper/src/encoder/mir/pure/specifications/encoder_poly.rs [hunk 21]
```diff
@@ -127,6 +127,31 @@ pub(super) fn inline_spec_item<'tcx>(
         .replace_multiple_places(&body_replacements))
 }
 
+fn error_ctxt(resource_type: &vir_crate::polymorphic::ResourceType) -> ErrorCtxt {
```
Comment (Aurel300) [inline prusti-viper/src/encoder/mir/pure/specifications/encoder_poly.rs]: This function specifically returns the error context related to time resources, so either rename it to something that includes `time`, move it inside `encode_time_specifications`, or even remove it (is it only used once).

File: prusti-viper/src/encoder/procedure_encoder.rs [hunk 22]
```diff
@@ -934,8 +940,11 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
         }
         // We'll add later more statements at the end of inv_pre_block, to havoc local variables
         let fnspec_span = {
-            let (stmts, fnspec_span) =
+            let (mut stmts, fnspec_span) =
                 self.encode_loop_invariant_inhale_fnspec_stmts(loop_head, before_invariant_block, false)?;
+            if config::time_reasoning() {
+                stmts.extend(self.get_tick_call(fnspec_span.clone(), 1).into_iter());
```
Comment (Aurel300) [inline prusti-viper/src/encoder/procedure_encoder.rs]: Add a comment explaining this `tick` call as well.

File: prusti-viper/src/encoder/procedure_encoder.rs [hunk 23]
```diff
@@ -7008,7 +7017,7 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
         })
     }
 
-    fn register_error<T: Into<MultiSpan>>(&self, span: T, error_ctxt: ErrorCtxt) -> vir::Position {
+    pub fn register_error<T: Into<MultiSpan>>(&self, span: T, error_ctxt: ErrorCtxt) -> vir::Position {
```
Comment (Aurel300) [inline prusti-viper/src/encoder/procedure_encoder.rs]: Where is this called from?

File: prusti-tests/tests/verify/fail/time-resources/array_cmp.rs [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+    if a[i] != b[i] {
+      return false;
+    }
+    i += 1;
+  }
+  return true;
+}
+
+#[requires(time_credits(1))]
+fn main() {}
```
Comment (Aurel300) [inline prusti-tests/tests/verify/fail/time-resources/array_cmp.rs]: (Make sure to end files with a line break.)

File: prusti-tests/tests/verify/fail/time-resources/variable.rs [hunk 25]
```diff
@@ -0,0 +1,11 @@
+// compile-flags: -Ptime_reasoning=true
+
+use prusti_contracts::*;
+
+// ignore-test: prusti-constract functions used outside specification check not yet implemented
```
Comment (Aurel300) [inline prusti-tests/tests/verify/fail/time-resources/variable.rs:5]: Remove this test.

File: prusti-tests/tests/verify/pass/time-resources/array_cmp.rs [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+
+use prusti_contracts::*;
+
+#[requires(a.len() == b.len())]
+#[requires(time_credits(a.len() + 1))]
+#[ensures(time_receipts(a.len() + 1))]
+fn safe_array_cmp<T: Eq>(a: &[T], b: &[T]) -> bool {
+  let mut i = 0;
+  let mut res = true;
+  while i < a.len() {
```
Comment (Aurel300) [inline prusti-tests/tests/verify/pass/time-resources/array_cmp.rs]: Why don't we need to put `time_credits` in the invariant?

File: prusti-tests/tests/verify/pass/time-resources/array_cmp.rs [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+  return res;
+}
+
+#[requires(a.len() == b.len())]
+#[requires(time_credits(a.len() + 1))]
+#[ensures(time_receipts(1))]
+fn opt_array_cmp<T: Eq>(a: &[T], b: &[T]) -> bool {
+  let mut i = 0;
+  while i < a.len() {
+    if a[i] != b[i] {
```
Comment (Aurel300) [inline prusti-tests/tests/verify/pass/time-resources/array_cmp.rs]: No body invariant?

File: prusti-tests/tests/verify/pass/time-resources/flag.rs [hunk 28]
```diff
@@ -0,0 +1,18 @@
+// compile-flags: -Ptime_reasoning=false
+
+use prusti_contracts::*;
+
+// This test checks that setting the time reasoning option to false ignores timing constrains
```
Comment (Aurel300) [inline prusti-tests/tests/verify/pass/time-resources/flag.rs]: As written in another comment, I don't think this should be the behaviour. This test also isn't even fully testing what the comment says: there is no timing specification to "ignore", only the default zero permissions.

File: prusti-tests/tests/verify/pass/time-resources/rec.rs [hunk 29]
```diff
@@ -0,0 +1,25 @@
+// compile-flags: -Ptime_reasoning=true
+
+use prusti_contracts::*;
+
+#[extern_spec]
+impl usize {
+    #[pure]
+    #[ensures(self.pow(n) * self == self.pow(n + 1))]
```
Comment (Aurel300) [inline prusti-tests/tests/verify/pass/time-resources/rec.rs:8]: I'm surprised this isn't running into issues due to cycles. Can you also call `fib` from `main` with a concrete argument?

File: prusti-common/src/vir/optimizations/functions/simplifier.rs (hunk 1 above)
Comment (Pialex99) [reply prusti-common/src/vir/optimizations/functions/simplifier.rs:60]: Not if `pos` is the default position which can happen if Prusti had inserted this boolean constant.

File: prusti-tests/tests/typecheck/fail/time-resources/condition.rs (hunk 8 above)
Comment (Pialex99) [reply prusti-tests/tests/typecheck/fail/time-resources/condition.rs:5]: As they don't make sense in executable code why don't we report them when we find them?

File: prusti-viper/src/encoder/procedure_encoder.rs (hunk 23 above)
Comment (Pialex99) [reply prusti-viper/src/encoder/procedure_encoder.rs]: In the body of `get_tick_call` in `encoder/resources/interface.rs`

File: prusti-tests/tests/typecheck/fail/time-resources/condition.rs (hunk 8 above)
Comment (Aurel300) [reply prusti-tests/tests/typecheck/fail/time-resources/condition.rs:5]: Because nobody has implemented that feature yet...

File: prusti-tests/tests/verify/pass/time-resources/array_cmp.rs (hunk 26 above)
Comment (Pialex99) [reply prusti-tests/tests/verify/pass/time-resources/array_cmp.rs]: Prusti didn't complain so I thought I didn't have to and the `loop.rs` tests were passing without so I thought that it was okay but playing a bit with the generated Viper code showed me that it doesn't 😅

File: prusti-tests/tests/verify/pass/time-resources/rec.rs (hunk 29 above)
Comment (Pialex99) [reply prusti-tests/tests/verify/pass/time-resources/rec.rs:8]: By adding `#[ensures(self >= 1 ==> self.pow(0) == 1)]`, yes I can use it in main.

File: prusti-viper/src/encoder/foldunfold/requirements.rs (hunk 20 above)
Comment (Pialex99) [reply prusti-viper/src/encoder/foldunfold/requirements.rs]: This modification was when we had the call to the `tick` method, but as we don't have it anymore we can remove those changes.

File: prusti-common/src/vir/optimizations/functions/simplifier.rs (hunk 2 above)
Comment (vakaras) [reply prusti-common/src/vir/optimizations/functions/simplifier.rs:98]: Yes, this can be a problem. For this and other reasons, it is better to use domain functions as triggers (with an axiom that says its value is always `true`).

**Comment by Aurel300** [comment]
Superseded by #1408.
